channeldb/migration14/migration.go

```diff
@@ -0,0 +1,98 @@
+package migration14
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+// MigrateCommitDiff migrates the CommitDiff struct on-disk for all active
+// channels to store the CommitSig with a length prefix, just as we do on the
+// wire. This ensures we'll be able to store any extra data appended to the
+// message, yet still serialize items after the message as well on disk.
+func MigrateCommitDiff(tx kvdb.RwTx) error {
+	openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+	if openChanBucket == nil {
+		return nil
+	}
+
+	return openChanBucket.ForEach(func(nodePub, v []byte) error {
+		// Ensure that this is a key the same size as a pubkey,
+		// and also that it leads directly to a bucket.
+		if len(nodePub) != 33 || v != nil {
+			return nil
+		}
+
+		nodeChanBucket := openChanBucket.NestedReadWriteBucket(nodePub)
+		if nodeChanBucket == nil {
+			return fmt.Errorf("no bucket for node %x", nodePub)
+		}
+
+		// The next layer down is all the chains that this node
+		// has channels on with us.
+		return nodeChanBucket.ForEach(func(chainHash, v []byte) error {
+			// If there's a value, it's not a bucket so
+			// ignore it.
+			if v != nil {
+				return nil
+			}
+
+			chainBucket := nodeChanBucket.NestedReadWriteBucket(
+				chainHash,
+			)
+			if chainBucket == nil {
+				return fmt.Errorf("unable to read "+
+					"bucket for chain=%x", chainHash)
+			}
+
+			return chainBucket.ForEach(func(chanPoint, v []byte) error {
+				// If there's a value, it's not a bucket so
+				// ignore it.
+				if v != nil {
+					return nil
+				}
+
+				// Once we've found a valid channel bucket,
+				// we'll extract it from the node's chain
+				// bucket.
+				chanBucket := chainBucket.NestedReadWriteBucket(
+					chanPoint,
+				)
+				if chanBucket == nil {
+					return fmt.Errorf("not bucket for "+
+						"channel: %x", chanPoint)
+				}
+
+				// At this point, we have the channel bucket
+				// now, so we'll check to see if this channel
+				// has a pending commitment or not. If it
+				// doesn't, then we can go on as there's
+				// nothing to migrate.
+				commitDiffBytes := chanBucket.Get(commitDiffKey)
+				if commitDiffBytes == nil {
+					return nil
+				}
+
+				// Now that we have the commit diff in the
+				// _old_ encoding, we'll write it back to disk
+				// using the new encoding which has a length
+				// prefix in front of the CommitSig.
```

Comment:
We've discussed before the pros of doing more low-level migrations. Can this be done here as well? Copy the bytes up to the message index, add a length byte, then copy the remaining bytes. Would minimize the dependencies needed and the amount of code to copy.

Reply:
I'm not too worried about deps or the amount of code. The code is really just dead code we put in there to gain confidence in the migration itself. The deps can be frozen using modules if we choose as well. Personally I can't really verify the low-level migrations during review without _re-generating_ all the bytes myself, so just more manual work, vs being able to read _deterministic_ code.

Reply:
> Copy the bytes up to the message index, add a length byte, then copy the remaining bytes

The amount of bytes preceding the message can't easily be deduced as we write the channel commitment before the diff, and the commitment contains elements with a variable size (the amount of HTLCs). The commitment itself isn't written with any sort of length prefix either.

Reply:
> > Copy the bytes up to the message index, add a length byte, then copy the remaining bytes
> 
> The amount of bytes preceding the message can't easily be deduced as we write the channel commitment before the diff, and the commitment contains elements with a variable size (the amount of HTLCs). The commitment itself isn't written with any sort of length prefix either.

Mhm, agree that that makes it tricky.